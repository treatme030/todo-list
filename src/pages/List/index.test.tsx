import React from "react";
import { Router, useLocation } from "react-router-dom";
import { createMemoryHistory } from 'history';
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import 'jest-styled-components';
import { ToDoListProvider } from "contexts";
import { List } from "./index";

describe('<List />', () => {
  it('renders component correctly', () => {
    const history = createMemoryHistory();
    history.push('/');
    localStorage.setItem('ToDoList', '["ToDo 1", "ToDo 2", "ToDo 3"]');

    const { container } = render(
      <ToDoListProvider>
        <Router history={history}>
          <List />
        </Router>
      </ToDoListProvider>
    );

    const toDoItem1 = screen.getByText('ToDo 1');
    expect(toDoItem1).toBeInTheDocument();
    expect(toDoItem1.getAttribute('href')).toBe('/detail/0');

    const toDoItem2 = screen.getByText('ToDo 2');
    expect(toDoItem2).toBeInTheDocument();
    expect(toDoItem2.getAttribute('href')).toBe('/detail/1');

    const toDoItem3 = screen.getByText('ToDo 3');
    expect(toDoItem3).toBeInTheDocument();
    expect(toDoItem3.getAttribute('href')).toBe('/detail/2');

    expect(screen.getAllByText('삭제').length).toBe(3);

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
  it('deletes toDo item', () => {
    const history = createMemoryHistory();
    history.push('/');
    localStorage.setItem('ToDoList', '["ToDo 1", "ToDo 2", "ToDo 3"]');

    render(
      <ToDoListProvider>
        <Router history={history}>
          <List />
        </Router>
      </ToDoListProvider>
    ); 
    
    const toDoItem = screen.getByText('ToDo 2');
    expect(toDoItem).toBeInTheDocument();
    const deleteButtons = screen.getAllByText('삭제');

    userEvent.click(deleteButtons[1]);
    expect(toDoItem).not.toBeInTheDocument();
    expect(localStorage.getItem('ToDoList')).toBe('["ToDo 1","ToDo 3"]');
  });
  it('moves to detail page', () => {
    const history = createMemoryHistory();
    history.push('/');
    localStorage.setItem('ToDoList', '["ToDo 1", "ToDo 2", "ToDo 3"]');

    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return (
        <div>{pathname}</div>
      );
    };

    render(
      <ToDoListProvider>
        <Router history={history}>
          <TestComponent />
          <List />
        </Router>
      </ToDoListProvider>
    ); 

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();
    const toDoItem = screen.getByText('ToDo 2');
    expect(toDoItem.getAttribute('href')).toBe('/detail/1');

    userEvent.click(toDoItem);
    expect(url.textContent).toBe('/detail/1');
  });
  it('moves to add page', () => {
    const history = createMemoryHistory();
    history.push('/');

    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return (
        <div>{pathname}</div>
      );
    };

    render(
      <ToDoListProvider>
        <Router history={history}>
          <TestComponent />
          <List />
        </Router>
      </ToDoListProvider>
    ); 

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('+');
    userEvent.click(addButton);
    expect(url.textContent).toBe('/add');
  });
});