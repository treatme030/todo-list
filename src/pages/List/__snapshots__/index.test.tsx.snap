// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> renders component correctly 1`] = `
.c4 {
  text-align: center;
  background-color: #ff1744;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c4:hover {
  background-color: #f01440;
}

.c4:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c5 {
  color: #fff;
  font-size: 16px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 1px solid #bdbdbd;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 10px;
  padding: 10px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 16px;
  margin-right: 20px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0,2);
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  font-size: 20px;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 30px;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  background-color: #304ffe;
  box-shadow: 5px 5px 10px rgba(0,0,0,0,2);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:hover {
  background-color: #1e40ff;
}

.c6:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0,2);
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
      data-testid="toDoList"
    >
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/detail/0"
        >
          ToDo 1
        </a>
        <button
          class="c4"
        >
          <label
            class="c5"
          >
            삭제
          </label>
        </button>
      </li>
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/detail/1"
        >
          ToDo 2
        </a>
        <button
          class="c4"
        >
          <label
            class="c5"
          >
            삭제
          </label>
        </button>
      </li>
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/detail/2"
        >
          ToDo 3
        </a>
        <button
          class="c4"
        >
          <label
            class="c5"
          >
            삭제
          </label>
        </button>
      </li>
    </ul>
    <a
      class="c6"
      href="/add"
    >
      +
    </a>
  </div>
</div>
`;
